lua/barbar/groups.lua

```diff
@@ -0,0 +1,206 @@
+--
+-- groups.lua
+--
+
+local table_insert = table.insert
+local strcharpart = vim.fn.strcharpart --- @type function
+local strwidth = vim.api.nvim_strwidth --- @type function
+
+local utils = require'barbar.utils'
+
```

Comment:
Missing `@class` and `@field` annotations, which will make type checking for the LSP hard since it has to assume types, rather than being able to verify what we've told it is true. Doing `require` in another file may not resolve function signatures without `@class` either.

Instead of using `local m`, the other files use `local <module name>` so that further declarations look like:

```lua
--- this
function render.update()
end

--- vs this
function m.update()
end
```

This makes it congruent with how the module will appear in other files.

Reply:
Ok for `@class`, seems to be enough for the LS to pick up the rest of the annotations.

Don't like `local groups`, we have params named `groups` for every function in there. We can refactor further in a subsequent PR.